File: demo_apps/android/app/src/main/java/com/demo/pdfviewer/MockSdk.kt
package com.demo.pdfviewer

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import kotlinx.coroutines.delay

class MockSdk {
    private var isOpened = false
    private var isModified = false

    suspend fun open(filePath: String): Boolean {
        // Simulate SDK initialization and parsing
        delay(800)
        isOpened = true
        isModified = false
        return true
    }

    suspend fun renderPage(pageIndex: Int): Bitmap {
        if (!isOpened) throw IllegalStateException("PDF not opened")
        // Simulate rendering time
        delay(300)

        // Return a mock Bitmap representing a PDF page
        val width = 800
        val height = 1100
        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)

        val paint = Paint()
        paint.color = Color.WHITE
        canvas.drawRect(0f, 0f, width.toFloat(), height.toFloat(), paint)

        paint.color = Color.LTGRAY
        paint.style = Paint.Style.STROKE
        paint.strokeWidth = 10f
        canvas.drawRect(0f, 0f, width.toFloat(), height.toFloat(), paint)

        paint.style = Paint.Style.FILL
        paint.color = Color.DKGRAY
        paint.textSize = 60f
        paint.textAlign = Paint.Align.CENTER
        canvas.drawText("Mock PDF Page ${pageIndex + 1}", width / 2f, height / 2f, paint)

        if (isModified) {
            paint.color = Color.BLUE
            paint.textSize = 40f
            canvas.drawText("(Modified)", width / 2f, height / 2f + 80f, paint)
        }

        return bitmap
    }

    suspend fun replaceText(searchString: String, replacementString: String): Int {
        if (!isOpened) throw IllegalStateException("PDF not opened")
        // Simulate text replacement processing
        delay(500)
        isModified = true
        return 1
    }

    suspend fun save(): Boolean {
        if (!isOpened) throw IllegalStateException("PDF not opened")
        // Simulate saving
        delay(1000)
        isModified = false
        return true
    }

    fun getPageCount(): Int {
        return if (isOpened) 5 else 0
    }
}

File: demo_apps/web/src/lib/mock-sdk.ts
export class PdfDocumentMock {
  private documentId: string | null = null;
  private isModified: boolean = false;

  async open(fileBuffer: ArrayBuffer): Promise<boolean> {
    console.log("PdfDocumentMock: Opening PDF document of size", fileBuffer.byteLength);
    // Simulate loading time
    await new Promise(resolve => setTimeout(resolve, 800));
    this.documentId = "mock-doc-123";
    this.isModified = false;
    return true;
  }

  async renderPage(pageIndex: number): Promise<string> {
    console.log(`PdfDocumentMock: Rendering page ${pageIndex}`);
    if (!this.documentId) {
      throw new Error("No document opened");
    }
    // Return a base64 encoded transparent 1x1 png or simple svg placeholder for the mock
    // Here we'll return a data URI for an SVG showing it's a rendered page
    const svg = `
      <svg width="800" height="1100" xmlns="http://www.w3.org/2000/svg">
        <rect width="100%" height="100%" fill="white" />
        <rect width="100%" height="100%" fill="none" stroke="#e5e7eb" stroke-width="4"/>
        <text x="50%" y="50%" font-family="Arial" font-size="40" fill="#9ca3af" text-anchor="middle" dominant-baseline="middle">
          Mock PDF Page ${pageIndex + 1}
        </text>
        ${this.isModified ? `<text x="50%" y="60%" font-family="Arial" font-size="30" fill="#3b82f6" text-anchor="middle" dominant-baseline="middle">(Modified)</text>` : ''}
      </svg>
    `;
    return `data:image/svg+xml;base64,${btoa(svg)}`;
  }

  async replaceText(searchString: string, replacementString: string): Promise<number> {
    console.log(`PdfDocumentMock: Replacing "${searchString}" with "${replacementString}"`);
    if (!this.documentId) {
      throw new Error("No document opened");
    }
    await new Promise(resolve => setTimeout(resolve, 500));
    this.isModified = true;
    // Return number of replacements
    return 1;
  }

  async save(): Promise<ArrayBuffer> {
    console.log("PdfDocumentMock: Saving PDF document");
    if (!this.documentId) {
      throw new Error("No document opened");
    }
    await new Promise(resolve => setTimeout(resolve, 1000));
    this.isModified = false;
    // Return a dummy ArrayBuffer
    return new ArrayBuffer(1024);
  }

  getPageCount(): number {
    return this.documentId ? 5 : 0;
  }
}

// Singleton instance to be used across the app
export const mockSdk = new PdfDocumentMock();
